src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,215 @@
 #include <cassert>
 #include <cstdint>
+#include <mutex>
+#include <optional>
+#include <unordered_map>
+#include <vector>
 
 #include <pthread.h>
 #include <sched.h>
 
 #include "affinity.h"
+#include "log.h"
 
-namespace celerity {
-namespace detail {
+namespace {
 
-	uint32_t affinity_cores_available() {
-		cpu_set_t available_cores;
-		[[maybe_unused]] const auto ret = pthread_getaffinity_np(pthread_self(), sizeof(cpu_set_t), &available_cores);
-		assert(ret == 0 && "Error retrieving affinity mask.");
-		return CPU_COUNT(&available_cores);
+using namespace celerity::detail::thread_pinning;
+
+std::vector<uint32_t> get_available_sequential_cores(const cpu_set_t& available_cores, const uint32_t count, const uint32_t starting_from_core) {
```

Comment:
Don't we want sequential as in next to each other? As far as I can tell this function allows for gaps?

Reply:
It's as sequential as we can make it. It's not necessarily *contiguous*, but it's also not called that ;)

Reply:
Hmm, ok. It would probably be preferable to find a contiguous range, but in the end the whole thing is a simple heuristic anyway. Speaking of: We should probably include a comment here that this function assumes that cores that are close together in the cpu set are also close in terms of cache, that we currently can't properly handle hyperthreads and that doing anything more fancy would require something like hwloc!

Reply:
I think the function doesn't really need that comment, the fact that we use it as the "auto" core pinning strategy does. But we have a comment to that effect already there.